README.md
# NFT DID Resolver

> NFT is a DID method that uses the Ceramic network to resolve DID documents for NFTs

> See [CIP-94](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-94/CIP-94.md)

## Getting started

This implementation is still a prototype. Contributions are welcome!

To use a package, you would need to provide three subgraph endpoints for every network you are going to use:
one for blocks, one for ERC721 tokens, another for ERC1155 tokens. You would also need to provide a `skew` that
is a time (in milliseconds) within which a latest block is considered valid. Usually it is a typical block time.

### Installation

```
$ npm install nft-did-resolver
```

### Usage

```typescript
import NftResolver, { NftResolverConfig } from 'nft-did-resolver'
import { Resolver } from 'did-resolver'
import Ceramic from '@ceramicnetwork/http-client'

const ceramic = new Ceramic() // connects to localhost:7007 by default

const config: NftResolverConfig = {
  ceramic,
  chains: {
    'eip155:1': {
      blocks: 'https://api.thegraph.com/subgraphs/name/yyong1010/ethereumblocks',
      skew: 15000,
      assets: {
        erc721: 'https://api.thegraph.com/subgraphs/name/sunguru98/mainnet-erc721-subgraph',
        erc1155: 'https://api.thegraph.com/subgraphs/name/sunguru98/mainnet-erc1155-subgraph',
      },
    },
    'eip155:4': {
      blocks: 'https://api.thegraph.com/subgraphs/name/mul53/rinkeby-blocks',
      skew: 15000,
      assets: {
        erc721: 'https://api.thegraph.com/subgraphs/name/sunguru98/erc721-rinkeby-subgraph',
        erc1155: 'https://api.thegraph.com/subgraphs/name/sunguru98/erc1155-rinkeby-subgraph',
      },
    },
  },
}

// getResolver will return an object with a key/value pair of { 'nft': resolver }
// where resolver is a function used by the generic did resolver.
const nftResolver = NftResolver.getResolver(config)
const didResolver = Resolver(nftResolver)

const erc721result = await didResolver.resolve(
  'did:nft:eip155:1_erc721:0xb300a43751601bd54ffee7de35929537b28e1488_2'
)
const erc1155result = await didResolver.resolve(
  'did:nft:eip155:1_erc1155:0x06eb48572a2ef9a3b230d69ca731330793b65bdc_1'
)
console.log(erc721result, erc1155result)
```

`chains` field in config has [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) chain identifiers as keys.
Each such `chain` is expected to contain endpoints to ERC721 and/or ERC1155 subgraphs under `assets` field.
Both ERC721 and ERC1155 are supported. Feel free to specify either one or both.

The resolver supports the following networks by default:

- Ethereum mainnet (`eip155:1`),
- Ethereum Rinkeby (`eip155:4`),
- Polygon (formerly Matic) (`eip155:137`).

If you use one of those, you do not have to provide `chains` field.

## Testing

```
$ npm test
```

## Custom Subgraphs

You may specify custom subgraph URLs in the configuration object as shown above in [usage](#usage).

**Note**: custom subgraphs must conform to the below schemas at a _minimum_ for assets to be resolved properly.

**Note**: At the moment, only ERC721 and ERC1155 asset namespaces are supported. However, CAIP2 chains beside ETH,
for instance xDAI, with support for those namespaces _are_ supported, as long as the subgraph schema is the same.

### ERC721:

```gql
type Token @entity {
  id: ID!
  contract: TokenContract!
  owner: Owner!
  ...
}

type TokenContract @entity {
  id: ID!
  tokens: [Token!]! @derivedFrom(field: "contract")
  ...
}

type Owner @entity {
  id: ID!
  tokens: [Token!]! @derivedFrom(field: "owner")
  ...
}

```

### ERC1155:

```gql
type Account @entity {
  id: ID!
  balances: [Balance!]! @derivedFrom(field: "account")
  ...
}

type TokenRegistry @entity {
  id: ID!
  tokens: [Token!]! @derivedFrom(field: "registry")
  ...
}

type Token @entity {
  id: ID!
  registry: TokenRegistry!
  identifier: BigInt!
  balances: [Balance!]! @derivedFrom(field: "token")
  ...
}

type Balance @entity {
  id: ID!
  token: Token!
  account: Account!
  ...
}

```

For more information on writing schemas for GraphProtocol, check out [their documentation](https://thegraph.com/docs/define-a-subgraph#defining-entities).

## DID Specs

The token DIDs are prefixed with `did:nft:`, and the latter half is a modified CAIP format.

**ERC721** ([CAIP-22](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/CAIP-22.md))

DID: `did:nft:{chainNamespace}:{chainReference}_erc721:{contractAddress}_{tokenId}`

CAIP-22: `{chainNamespace}:{chainReference}/erc721:{contractAddress}/{tokenId}`

**ERC1155** ([CAIP-29](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/CAIP-29.md))

DID: `did:nft:{chainNamespace}:{chainReference}_erc1155:{contractAddress}_{tokenId}`

CAIP-29: `{chainNamespace}:{chainReference}/erc1155:{contractAddress}/{tokenId}`

### Conversions

**DID->CAIP**

```
const caip = did.substr(8).replace(/_/g, '/')
```

**CAIP->DID**

```
const did = `did:nft:${caip.replace(/\//g, '_')
```

There are helpers that help you with the conversion:

```typescript
import { caipToDid, didToCaip, createNftDidUrl } from 'nft-did-resolver'
import { AssetId } from 'caip'

// CAIP -> DID URL
const didUrl = createNftDidUrl({
  chainId: 'eip155:1',
  namespace: 'erc721:0x1234567891234567891234567891234596351156',
  tokenId: '1',
})
// If you use `caip` library in your app, consider using sister `caipToDid` function to convert `AssetId` to NFT DID URL. 

// DID URL -> CAIP
const assetId1 = didToCaip(didUrl) // eip155:1/erc721:0x1234567891234567891234567891234596351156/1
const assetId2 = didToCaip(didUrlWithTimestamp) // eip155:1/erc721:0x1234567891234567891234567891234596351156/1
```

## Contributing

We are happy to accept small and large contributions. Make sure to check out the [Ceramic specifications](https://github.com/ceramicnetwork/specs) for details of how the protocol works.

## License

Apache-2.0 OR MIT


CHANGELOG.md
# Change Log

## v0.2.1 - 2021-08-12

Use GA `caip` library, not beta one.

## v0.2.0 - 2021-08-07

First GA release of `nft-did-resolver`.

## v0.1.0 - 2021-04-14
First release of `nft-did-resolver` 🎉



src/__tests__/nft-did-vector.ts
import { DIDResolutionResult, VerificationMethod } from 'did-resolver'

export class NftDidVector {
  nftDid: string
  nftOwners: string[] | undefined
  verificationMethods: VerificationMethod[] | undefined
  versionId: string | undefined
  versionTime: string | undefined
  caip10Controller: string | undefined
  errorMessage: string | undefined

  constructor(vectorBuilder: NftDidVectorBuilder) {
    this.nftDid = vectorBuilder.nftDid
    this.nftOwners = vectorBuilder.nftOwners
    this.verificationMethods = vectorBuilder.verificationMethods
    this.versionId = vectorBuilder.versionId
    this.versionTime = vectorBuilder.versionTime
    this.caip10Controller = vectorBuilder.caip10Controller
    this.errorMessage = vectorBuilder.errorMessage
  }

  getDidWithVersionId(): string {
    return this.nftDid.concat(`?versionId=${this.versionId}`)
  }

  getDidWithVersionTime(): string {
    return this.nftDid.concat(`?versionTime=${this.versionTime}`)
  }

  getResult(): DIDResolutionResult {
    if (this.errorMessage) {
      return {
        didDocument: null,
        didDocumentMetadata: {},
        didResolutionMetadata: {
          error: 'invalidDid',
          message: this.errorMessage,
        },
      }
    }

    const resolutionResult = {
      didDocument: {
        id: this.nftDid,
      },
      didDocumentMetadata: {},
      didResolutionMetadata: { contentType: 'application/did+json' },
    } as DIDResolutionResult

    if (this.verificationMethods)
      resolutionResult.didDocument.verificationMethod = [...this.verificationMethods]
    if (this.caip10Controller) resolutionResult.didDocument.controller = this.caip10Controller
    return resolutionResult
  }
}

export class NftDidVectorBuilder {
  public readonly nftNamespace: string
  public readonly caip2ChainId: string

  public nftDid = '' // falsey, will throw if not made or provided
  public nftContract: string | undefined
  public nftId: string | undefined
  public nftOwners: string[] | undefined

  public verificationMethods: VerificationMethod[] | undefined
  public versionId: string | undefined
  public versionTime: string | undefined
  public caip10Controller: string | undefined

  public errorMessage: string | undefined

  constructor(caip2ChainId: string, nftNamespace: string) {
    this.caip2ChainId = caip2ChainId
    this.nftNamespace = nftNamespace
  }

  setNftContract(nftContract: string): NftDidVectorBuilder {
    this.nftContract = nftContract
    return this
  }

  setNftId(nftId: string): NftDidVectorBuilder {
    this.nftId = nftId
    return this
  }

  setNftOwners(nftOwners: string[]): NftDidVectorBuilder {
    this.nftOwners = nftOwners
    return this
  }

  setNftDid(nftDid: string): NftDidVectorBuilder {
    this.nftDid = nftDid
    return this
  }

  setVerificationMethods(methods: VerificationMethod[]): NftDidVectorBuilder {
    this.verificationMethods = methods
    return this
  }

  setCaip10Controller(caip10Controller: string): NftDidVectorBuilder {
    this.caip10Controller = caip10Controller
    return this
  }

  setErrorMessage(errorMessage: string): NftDidVectorBuilder {
    this.errorMessage = errorMessage
    return this
  }

  setVersionId(versionId: string): NftDidVectorBuilder {
    this.versionId = versionId
    return this
  }

  // Should be ISOString
  setVersionTime(versionTime: string): NftDidVectorBuilder {
    this.versionTime = versionTime
    return this
  }

  build(): NftDidVector {
    if (!this.nftDid) this.nftDid = this.makeDid()

    if (!this.errorMessage && !this.verificationMethods)
      this.verificationMethods = this.makeVerificationMethods()

    return new NftDidVector(this)
  }

  private makeDid(): string {
    if (!this.nftContract || !this.nftId) {
      throw new Error('Must provide contract address and id OR DID.')
    }
    return `did:nft:${this.caip2ChainId}_${this.nftNamespace}:${this.nftContract}_${this.nftId}`
  }

  private makeVerificationMethods(): VerificationMethod[] {
    if (!this.nftOwners) {
      throw new Error('Must provide NftOwners')
    } else if (!this.nftDid) {
      throw new Error('Must provide Nft DID or args')
    }

    return this.nftOwners.slice().map((owner) => {
      return {
        id: `${this.nftDid}#${owner.slice(2, 14)}`,
        type: 'BlockchainVerificationMethod2021',
        controller: this.nftDid,
        blockchainAccountId: `${this.caip2ChainId}:${owner}`,
      } as VerificationMethod
    })
  }
}


src/subgraphUtils.ts
import fetch from 'cross-fetch'
import { jsonToGraphQLQuery } from 'json-to-graphql-query'
import { AssetID } from '.'

const GRAPH_API_PREFIX = 'https://api.thegraph.com/subgraphs/name'
const BLOCK_QUERY_URL = `${GRAPH_API_PREFIX}/yyong1010/ethereumblocks`
const ERC721_QUERY_URL = `${GRAPH_API_PREFIX}/wighawag/eip721-subgraph`
const ERC1155_QUERY_URL = `${GRAPH_API_PREFIX}/amxx/eip1155-subgraph`

export const fetchQueryData = async (queryUrl: string, query: unknown): Promise<any> => {
  const fetchOpts = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query: jsonToGraphQLQuery({ query }) }),
  }

  const resp = await fetch(queryUrl, fetchOpts)

  if (resp.ok) {
    const { data, error } = await resp.json()
    if (error) throw new Error(error.message)
    return data
  } else {
    throw new Error('Received an invalid response from TheGraph API')
  }
}

type BlockQueryResponse = {
  blocks: {
    number: string
  }[]
}

/**
 * Queries TheGraph to find the latest block at the given time.
 * @param timestamp
 * @returns {string} latest block num at timestamp
 */
export const blockAtTime = async (timestamp: number): Promise<number> => {
  const query = {
    blocks: {
      __args: {
        first: 1,
        orderBy: 'timestamp',
        orderDirection: 'desc',
        where: {
          // we ask for lte because it is the last known block at the given time
          timestamp_lte: timestamp,
        },
      },
      number: true,
    },
  }

  const queryData = (await fetchQueryData(BLOCK_QUERY_URL, query)) as BlockQueryResponse

  if (!queryData?.blocks) {
    throw new Error('Missing data from subgraph query')
  } else if (queryData.blocks.length === 0) {
    throw new Error(`No blocks exist before timestamp: ${timestamp}`)
  }

  return parseInt(queryData.blocks[0].number)
}
/**
 * Eth blocks are typically 13 seconds. We use this check so we don't have to
 * make an unneccessary call to the blocks subgraph if the did was just created.
 */
export const isWithinLastBlock = (timestamp: number): boolean => {
  return Date.now() - timestamp <= 10 * 1000
}

type ERC721DataResponse = {
  tokens: {
    owner: {
      id: string
    }
  }[]
}

export const erc721OwnerOf = async (
  asset: AssetID,
  blockNum: number,
  customSubgraph?: string
): Promise<string> => {
  const query = {
    tokens: {
      __args: {
        where: {
          // contract: asset.reference, // not necessary
          id: [asset.reference, asset.tokenId].join('_'),
        },
        first: 1,
        block: blockNum ? { number: blockNum } : null,
      },
      owner: {
        id: true,
      },
    },
  }

  const queryUrl = customSubgraph || ERC721_QUERY_URL
  const queryData = (await fetchQueryData(queryUrl, query)) as ERC721DataResponse

  if (!queryData?.tokens) {
    throw new Error('Missing data from subgraph query')
  } else if (queryData.tokens.length === 0) {
    throw new Error(
      `No owner found for ERC721 NFT ID: ${asset.tokenId} for contract: ${asset.reference}`
    )
  }

  return queryData.tokens[0].owner.id
}

type ERC1155DataResponse = {
  tokens: {
    balances: {
      account: {
        id: string
      }
    }[]
  }[]
}

export const erc1155OwnersOf = async (
  asset: AssetID,
  blockNum: number,
  customSubgraph?: string
): Promise<string[]> => {
  const query = {
    tokens: {
      __args: {
        where: {
          registry: asset.reference,
          identifier: asset.tokenId,
          // id: [asset.reference, `0x${asset.tokenId}`].join('-') // could use this instead
        },
        first: 1,
        block: blockNum ? { number: blockNum } : null,
      },
      balances: {
        __args: {
          where: {
            value_gt: 0, // we don't want to get an "owner" with a balance of zero
          },
        },
        account: {
          id: true,
        },
      },
    },
  }

  const queryUrl = customSubgraph || ERC1155_QUERY_URL
  const queryData = (await fetchQueryData(queryUrl, query)) as ERC1155DataResponse

  if (!queryData?.tokens[0]) {
    throw new Error(
      `No tokens with ERC1155 NFT ID: ${asset.tokenId} found for contract: ${asset.reference}`
    )
  } else if (!queryData.tokens[0].balances || queryData.tokens[0].balances.length === 0) {
    throw new Error(
      `No owner found for ERC1155 NFT ID: ${asset.tokenId} for contract: ${asset.reference}`
    )
  }

  return queryData.tokens[0].balances.slice().map((bal) => bal.account.id)
}


src/subgraph-utils.ts
import fetch from 'cross-fetch'
import { jsonToGraphQLQuery } from 'json-to-graphql-query'
import { AssetId } from 'caip'
import BigNumber from 'bignumber.js'

export const fetchQueryData = async (queryUrl: string, query: unknown): Promise<any> => {
  const fetchOpts = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query: jsonToGraphQLQuery({ query }) }),
  }

  const resp = await fetch(queryUrl, fetchOpts)

  if (resp.ok) {
    const { data, error } = await resp.json()
    if (error) throw new Error(error.message)
    return data
  } else {
    throw new Error('Received an invalid response from TheGraph API')
  }
}

type BlockQueryResponse = {
  blocks: {
    number: string
  }[]
}

/**
 * Queries TheGraph to find the latest block at the given time.
 * @param timestamp
 * @param blockQueryUrl - subgraph url for blocks
 * @returns {string} latest block num at timestamp
 */
export const blockAtTime = async (timestamp: number, blockQueryUrl: string): Promise<number> => {
  const query = {
    blocks: {
      __args: {
        first: 1,
        orderBy: 'timestamp',
        orderDirection: 'desc',
        where: {
          // we ask for lte because it is the last known block at the given time
          timestamp_lte: timestamp,
        },
      },
      number: true,
    },
  }

  const queryData = (await fetchQueryData(blockQueryUrl, query)) as BlockQueryResponse

  if (!queryData?.blocks) {
    throw new Error('Missing data from subgraph query')
  } else if (queryData.blocks.length === 0) {
    throw new Error(`No blocks exist before timestamp: ${timestamp}`)
  }

  return parseInt(queryData.blocks[0].number)
}
/**
 * Eth blocks are typically 13 seconds. We use this check so we don't have to
 * make an unneccessary call to the blocks subgraph if the did was just created.
 */
export const isWithinLastBlock = (timestamp: number, skew: number): boolean => {
  return Date.now() - timestamp <= skew
}

type ERC721DataResponse = {
  tokens: {
    owner: {
      id: string
    }
  }[]
}

export const erc721OwnerOf = async (
  asset: AssetId,
  blockNum: number,
  queryUrl: string
): Promise<string> => {
  const tokenId = `0x${new BigNumber(asset.tokenId).toString(16)}`
  const query = {
    tokens: {
      __args: {
        where: {
          id: [asset.assetName.reference, tokenId].join('-'),
        },
        first: 1,
        block: blockNum ? { number: blockNum } : null,
      },
      owner: {
        id: true,
      },
    },
  }

  const queryData = (await fetchQueryData(queryUrl, query)) as ERC721DataResponse

  if (!queryData?.tokens) {
    throw new Error('Missing data from subgraph query')
  } else if (queryData.tokens.length === 0) {
    throw new Error(
      `No owner found for ERC721 NFT ID: ${asset.tokenId} for contract: ${asset.assetName.reference}`
    )
  }

  return queryData.tokens[0].owner.id
}

type ERC1155DataResponse = {
  tokens: {
    balances: {
      account: {
        id: string
      }
    }[]
  }[]
}

export const erc1155OwnersOf = async (
  asset: AssetId,
  blockNum: number,
  queryUrl: string
): Promise<string[]> => {
  const tokenId = `0x${new BigNumber(asset.tokenId).toString(16)}`
  const query = {
    tokens: {
      __args: {
        where: {
          id: [asset.assetName.reference, tokenId].join('-'),
        },
        first: 1,
        block: blockNum ? { number: blockNum } : null,
      },
      balances: {
        __args: {
          where: {
            value_gt: 0, // we don't want to get an "owner" with a balance of zero
          },
        },
        account: {
          id: true,
        },
      },
    },
  }

  const queryData = (await fetchQueryData(queryUrl, query)) as ERC1155DataResponse

  if (!queryData?.tokens[0]) {
    throw new Error(
      `No tokens with ERC1155 NFT ID: ${asset.tokenId} found for contract: ${asset.assetName.reference}`
    )
  } else if (!queryData.tokens[0].balances || queryData.tokens[0].balances.length === 0) {
    throw new Error(
      `No owner found for ERC1155 NFT ID: ${asset.tokenId} for contract: ${asset.assetName.reference}`
    )
  }

  return queryData.tokens[0].balances.slice().map((bal) => bal.account.id)
}


src/index.ts
import type {
  DIDResolutionResult,
  DIDResolutionOptions,
  DIDDocument,
  ParsedDID,
  Resolver,
  ResolverRegistry,
  VerificationMethod,
} from 'did-resolver'
import type { CeramicApi } from '@ceramicnetwork/common'
import { Caip10Link } from '@ceramicnetwork/stream-caip10-link'
import { ChainId, AccountId, AssetId } from 'caip'
import { DIDDocumentMetadata } from 'did-resolver'
import { blockAtTime, erc1155OwnersOf, erc721OwnerOf, isWithinLastBlock } from './subgraph-utils'
import merge from 'merge-options'

const DID_LD_JSON = 'application/did+ld+json'
const DID_JSON = 'application/did+json'

export function didToCaip(id: string): AssetId {
  const caip = id
    .replace(/^did:nft:/, '')
    .replace(/\?.+$/, '')
    .replace(/_/g, '/')
  return new AssetId(AssetId.parse(caip))
}

async function assetToAccount(
  asset: AssetId,
  timestamp: number | undefined,
  chains: Record<string, ChainConfig | undefined>
): Promise<AccountId[]> {
  const assetChainId = asset.chainId.toString()
  const chain = chains[assetChainId]
  if (!chain) {
    throw new Error(`No chain configuration for ${assetChainId}`)
  }

  // we want to query what block is at the timestamp IFF it is an (older) existing timestamp
  let queryBlock = 0
  if (timestamp && !isWithinLastBlock(timestamp, chain.skew)) {
    queryBlock = await blockAtTime(timestamp, chain.blocks)
  }

  let owners: string[]
  let ercSubgraphUrls = undefined

  if (chains && chains[assetChainId]) {
    ercSubgraphUrls = chains[assetChainId].assets
  }

  if (asset.assetName.namespace === 'erc721') {
    owners = [await erc721OwnerOf(asset, queryBlock, ercSubgraphUrls?.erc721)]
  } else if (asset.assetName.namespace === 'erc1155') {
    owners = await erc1155OwnersOf(asset, queryBlock, ercSubgraphUrls?.erc1155)
  } else {
    throw new Error(
      `Only erc721 and erc1155 namespaces are currently supported. Given: ${asset.assetName.namespace}`
    )
  }

  return owners.slice().map(
    (owner) =>
      new AccountId({
        chainId: asset.chainId,
        address: owner,
      })
  )
}

/**
 * Creates CAIP-10 links for each account to be used as controllers.
 * Since there may be many owners for a given NFT (only ERC1155 for now),
 * there can be many controllers of that DID document.
 */
async function accountsToDids(
  accounts: AccountId[],
  timestamp: number,
  ceramic: CeramicApi
): Promise<string[] | undefined> {
  const controllers: string[] = []

  const links = await Promise.all(
    accounts.map((accountId: AccountId) => Caip10Link.fromAccount(ceramic, accountId))
  )

  for (const link of links) {
    if (link?.did) controllers.push(link.did)
  }

  return controllers.length > 0 ? controllers : undefined
}

function wrapDocument(did: string, accounts: AccountId[], controllers?: string[]): DIDDocument {
  // Each of the owning accounts is a verification method (at the point in time)
  const verificationMethods = accounts.slice().map<VerificationMethod>((account) => {
    return {
      id: `${did}#${account.address.slice(2, 14)}`,
      type: 'BlockchainVerificationMethod2021',
      controller: did,
      blockchainAccountId: account.toString(),
    }
  })

  const doc: DIDDocument = {
    id: did,
    verificationMethod: [...verificationMethods],
  }

  // Controllers should only be an array when there're more than one
  if (controllers) doc.controller = controllers.length === 1 ? controllers[0] : controllers

  return doc
}

/**
 * Gets the unix timestamp from the `versionTime` parameter.
 * @param query
 */
function getVersionTime(query = ''): number {
  const versionTime = query.split('&').find((e) => e.includes('versionTime'))
  if (versionTime) {
    return Math.floor(new Date(versionTime.split('=')[1]).getTime() / 1000)
  }
  return 0 // 0 is falsey
}

function validateResolverConfig(config: Partial<NftResolverConfig>) {
  if (!config) {
    throw new Error(`Missing nft-did-resolver config`)
  }
  if (!config.ceramic) {
    throw new Error('Missing ceramic client in nft-did-resolver config')
  }
  const chains = config.chains
  if (!chains) {
    throw new Error('Missing chain parameters in nft-did-resolver config')
  }
  try {
    Object.entries(config.chains).forEach(([chainId, chainConfig]) => {
      ChainId.parse(chainId)
      new URL(chainConfig.blocks)
      Object.values(chainConfig.assets).forEach((subgraph) => {
        new URL(subgraph)
      })
    })
  } catch (e) {
    throw new Error(`Invalid config for nft-did-resolver: ${e.message}`)
  }
}

export type ChainConfig = {
  blocks: string
  skew: number
  assets: Record<string, string>
}

/**
 * When passing in a custom subgraph url, it must conform to the same standards as
 * represented by the included ERC721 and ERC1155 subgraphs
 * Example:
 * ```
 * const customConfig = {
 *  ceramic: ceramicClient,
 *  chains: {
 *    "eip155:1": {
 *      "blocks": "https://api.thegraph.com/subgraphs/name/yyong1010/ethereumblocks",
 *      "skew": 15000, // in milliseconds
 *      "assets": {
 *        "erc1155": "https://api.thegraph.com/subgraphs/name/amxx/eip1155-subgraph",
 *        "erc721": "https://api.thegraph.com/subgraphs/name/touchain/erc721track",
 *      }
 *    }
 *  }
 * }
 * ```
 */
export type NftResolverConfig = {
  ceramic: CeramicApi
  chains: Record<string, ChainConfig>
}

async function resolve(
  did: string,
  methodId: string,
  timestamp: number,
  config: NftResolverConfig
): Promise<DIDResolutionResult> {
  const asset = didToCaip(methodId)
  // for 1155s, there can be many accounts that own a single asset
  const owningAccounts = await assetToAccount(asset, timestamp, config.chains)
  const controllers = await accountsToDids(owningAccounts, timestamp, config.ceramic)
  const metadata: DIDDocumentMetadata = {}

  // TODO create (if it stays in the spec)

  return {
    didResolutionMetadata: { contentType: DID_JSON },
    didDocument: wrapDocument(did, owningAccounts, controllers),
    didDocumentMetadata: metadata,
  }
}

/**
 * Convert AssetId to did:nft URL, including timestamp, if present.
 * @param assetId - NFT Asset
 * @param timestamp - JS Time as unix timestamp.
 */
export function caipToDid(assetId: AssetId, timestamp?: number): string {
  const query = timestamp
    ? `?versionTime=${new Date(timestamp * 1000).toISOString().split('.')[0] + 'Z'}`
    : ''

  const id = assetId.toString().replace(/\//g, '_')
  return `did:nft:${id}${query}`
}

function withDefaultConfig(config: Partial<NftResolverConfig>): NftResolverConfig {
  const defaults = {
    chains: {
      'eip155:1': {
        blocks: 'https://api.thegraph.com/subgraphs/name/yyong1010/ethereumblocks',
        skew: 15000,
        assets: {
          erc721: 'https://api.thegraph.com/subgraphs/name/sunguru98/mainnet-erc721-subgraph',
          erc1155: 'https://api.thegraph.com/subgraphs/name/sunguru98/mainnet-erc1155-subgraph',
        },
      },
      'eip155:4': {
        blocks: 'https://api.thegraph.com/subgraphs/name/mul53/rinkeby-blocks',
        skew: 15000,
        assets: {
          erc721: 'https://api.thegraph.com/subgraphs/name/sunguru98/erc721-rinkeby-subgraph',
          erc1155: 'https://api.thegraph.com/subgraphs/name/sunguru98/erc1155-rinkeby-subgraph',
        },
      },
      'eip155:137': {
        blocks: 'https://api.thegraph.com/subgraphs/name/matthewlilley/polygon-blocks',
        skew: 2200,
        assets: {
          erc721: 'https://api.thegraph.com/subgraphs/name/yellow-heart/maticsubgraph',
          erc1155: 'https://api.thegraph.com/subgraphs/name/tranchien2002/eip1155-matic',
        },
      },
    },
  }
  return merge.bind({ ignoreUndefined: true })(defaults, config)
}

/**
 * Params for `createNftDidUrl` function.
 */
export type NftDidUrlParams = {
  /**
   * CAIP-10 Chain ID. For example: `eip155:1`
   */
  chainId: string
  /**
   * Token namespace: `erc721` or `erc1155`
   */
  namespace: string,
  /**
   * Token contract address
   */
  contract: string
  /**
   * Token ID
   */
  tokenId: string
  /**
   * Unix timestamp for `versionTime` DID URL query param. Helps to find NFT owners at particular point in time.
   */
  timestamp?: number
}

/**
 * Convert NFT asset id to NFT DID URL. Can include timestamp (as unix timestamp) if provided.
 */
export function createNftDidUrl(params: NftDidUrlParams): string {
  if (!['erc721', 'erc1155'].includes(params.namespace)) {
    throw new Error(`Only erc721 and erc1155 are supported`)
  }
  return caipToDid(
    new AssetId({
      chainId: params.chainId,
      assetName: `${params.namespace}:${params.contract}`,
      tokenId: params.tokenId,
    }),
    params.timestamp
  )
}

export default {
  getResolver: (
    config: Partial<NftResolverConfig> & Required<{ ceramic: CeramicApi }>
  ): ResolverRegistry => {
    config = withDefaultConfig(config)
    validateResolverConfig(config)
    return {
      nft: async (
        did: string,
        parsed: ParsedDID,
        resolver: Resolver,
        options: DIDResolutionOptions
      ): Promise<DIDResolutionResult> => {
        const contentType = options.accept || DID_JSON
        try {
          const timestamp = getVersionTime(parsed.query)
          const didResult = await resolve(did, parsed.id, timestamp, config as NftResolverConfig)

          if (contentType === DID_LD_JSON) {
            didResult.didDocument['@context'] = 'https://w3id.org/did/v1'
            didResult.didResolutionMetadata.contentType = DID_LD_JSON
          } else if (contentType !== DID_JSON) {
            didResult.didDocument = null
            didResult.didDocumentMetadata = {}
            delete didResult.didResolutionMetadata.contentType
            didResult.didResolutionMetadata.error = 'representationNotSupported'
          }
          return didResult
        } catch (e) {
          return {
            didResolutionMetadata: {
              error: 'invalidDid',
              message: e.toString(),
            },
            didDocument: null,
            didDocumentMetadata: {},
          }
        }
      },
    }
  },
}
